packages/near-membrane-base/src/blue.ts

```diff
@@ -125,8 +107,19 @@ export function blueProxyFactory(env: MembraneBroker) {
         ObjectDefineProperties(shadowTarget, blueDescriptors);
     }
 
-    function createBlueProxy(red: BlueProxyTarget): BlueProxy {
-        const shadowTarget = createBlueShadowTarget(red);
+    function createBlueProxy(
+        red: BlueProxyTarget,
+        targetTypeof: ProxyTargetType,
+        targetIsArrowFunction: boolean,
+        targetFunctionName: string | undefined,
+        targetIsArray: boolean
```

Comment:
Do we think this list of args might grow? Would it make sense to convert this to an options object?

Reply:
@rwaldron 
> Do we think this list of args might grow? Would it make sense to convert this to an options object?

That's what a [bitmask](https://github.com/salesforce/near-membrane/pull/176#discussion_r685403143) is good for. Options as field of bits instead of objects (see [this](https://github.com/standard-things/esm/commit/cc46dc07b418752540a866ddefee4647d905a013#diff-f754260c59b6ebf07d562831935c5111e1ee3cbd684917e6ba6e5dfc0677fb46) for example).

Reply:
bitmask seems interesting but seems like something to have after compilation time?

Reply:
@leobalter 
> bitmask seems interesting but seems like something to have after compilation time?

This lib is the lowest level bit we have. Considerations about object creation, string comparisons, etc. come in to play. A bitmask can be seen as an alternative to options bags that are still semi-readable when used with descriptively named all-cap constants.

Reply:
bitmask reflects what is used in industry standards for memory and perf critical code, e.g. Moddable XS, so there is precedent if we eventually use it.

Reply:
> @rwaldron
> 
> > Do we think this list of args might grow? Would it make sense to convert this to an options object?
> 
> That's what a [bitmask](https://github.com/salesforce/near-membrane/pull/176#discussion_r685403143) is good for. Options as a field of bits instead of objects (see [this](https://github.com/standard-things/esm/commit/cc46dc07b418752540a866ddefee4647d905a013#diff-f754260c59b6ebf07d562831935c5111e1ee3cbd684917e6ba6e5dfc0677fb46) for example).

Now we're talking. I've long been an advocate for greater use of bitfields/bitmaps (the _bitmask_ is not the state itself, it's what you would use for performing bitwise operations on the state, ie. the composite value).

Reply:
The only issue I see with the data above is that a `targetFunctionName` value would have to be stored somewhere else. 

A hypothetical bitfield exercise: 

```js
const TARGET_FUNCTION_NAME_MASK = 0b1111_1111_1111_1111; // 0-16

// TYPEOF_*
const TARGET_TYPEOF_FUNCTION_INDEX = 0b0000_0001_0000_0000_0000_0000; // 17
const TARGET_TYPEOF_OBJECT_INDEX = 0b0000_0001_0000_0000_0000_0000; // 18

// IS_*
const TARGET_IS_ARRAY_INDEX = 0b0001_0000_0000_0000_0000_0000_0000; // 25
const TARGET_IS_ARROW_FUNCTION_INDEX = 0b0010_0000_0000_0000_0000_0000_0000; // 26
const TARGET_IS_ARROW_FUNCTION_MASK = TARGET_TYPEOF_FUNCTION_INDEX | TARGET_IS_ARROW_FUNCTION_INDEX;

// Validation
const TARGET_TYPEOF_INVALID = TARGET_TYPEOF_FUNCTION_INDEX | TARGET_TYPEOF_OBJECT_INDEX;
const TARGET_IS_INVALID = TARGET_IS_ARRAY_INDEX | TARGET_IS_ARROW_FUNCTION_INDEX;

// -------------------------------------

var flags = 0b0000_0000_0000_0000_0000_0000_0000_0000;

// Set the TARGET_IS_ARRAY_INDEX bit
flags |= TARGET_IS_ARRAY_INDEX << 1;

console.log(`target is an array? ${!!((flags >> 1) & TARGET_IS_ARRAY_INDEX)} (expecting: true)`);

// Toggle the TARGET_IS_ARRAY_INDEX bit
flags ^= TARGET_IS_ARRAY_INDEX << 1;

console.log(`TARGET_IS_ARRAY_INDEX toggled`);
console.log(`target is an array? ${!!((flags >> 1) & TARGET_IS_ARRAY_INDEX)} (expecting: false)`); 

flags |= TARGET_TYPEOF_FUNCTION_INDEX << 1;
flags |= TARGET_IS_ARROW_FUNCTION_INDEX << 1;

console.log(`target is an array? ${!!((flags >> 1) & TARGET_IS_ARRAY_INDEX)} (expecting: false)`); 
console.log(`target is a function? ${!!((flags >> 1) & TARGET_TYPEOF_FUNCTION_INDEX)} (expecting: true)`);
console.log(`target is an arrow function? ${!!((flags >> 1) & TARGET_IS_ARROW_FUNCTION_INDEX)} (expecting: true)`);
console.log(`target is an arrow function (mask)? ${!!(flags & TARGET_IS_ARROW_FUNCTION_MASK)} (expecting: true)`);


// Toggle the TARGET_IS_ARROW_FUNCTION_INDEX bit
flags ^= TARGET_IS_ARROW_FUNCTION_INDEX << 1;

console.log(`TARGET_IS_ARROW_FUNCTION_INDEX toggled`);

console.log(`target is an array? ${!!((flags >> 1) & TARGET_IS_ARRAY_INDEX)} (expecting: false)`); 
console.log(`target is a function? ${!!((flags >> 1) & TARGET_TYPEOF_FUNCTION_INDEX)} (expecting: true)`);
console.log(`target is an arrow function? ${!!((flags >> 1) & TARGET_IS_ARROW_FUNCTION_INDEX)} (expecting: false)`); 
console.log(`target is an arrow function (mask)? ${!!(flags & TARGET_IS_ARROW_FUNCTION_MASK)} (expecting: false)`); 
console.log(`target is a non-arrow function? ${!(flags & TARGET_IS_ARROW_FUNCTION_MASK)} (expecting: true)`);


const functionNameList = new Array();

function f0() {}
function f1() {}
function f2() {}
function f3() {}
function f4() {}
function f5() {}

functionNameList.push(f0.name);
functionNameList.push(f1.name);
functionNameList.push(f2.name);
functionNameList.push(f3.name);
functionNameList.push(f4.name);
functionNameList.push(f5.name);
  

// ... later ...
flags = 0b0000_0000_0000_0000_0000_0000_0000_0000;

// target is the non-arrow function "f3"
flags |= TARGET_TYPEOF_FUNCTION_INDEX << 1;
flags |= functionNameList.indexOf(f3.name);


// ... even later ...


const functionNameIndex = flags & TARGET_FUNCTION_NAME_MASK;

const functionNameValue = functionNameList[functionNameIndex];

console.log(`retrieved the correct name? ${functionNameValue === f3.name}`);

Reply:
I'm good with bitmasks, can't be an options bag because it is supposed to be used across a callable boundary that do not accept objects. Passing one argument with the bits is way better than passing 10 arguments with string comparison, so bitmasks are preferred. I can work on that, or someone else can... either way it seems that it is the preferred path forward.

Reply:
Ok, got a chance to add the bitwise stuff... please take another look. heads up: it is not super pretty, but it works fine.

Reply:
> I'm good with bitmasks, can't be an options bag because it is supposed to be used across a callable boundary that do not accept objects.

I hadn't even considered that, but it's a REALLY compelling point.

Reply:
updated... this now looks a lot clean.